relay-server/src/actors/spooler/mod.rs

```diff
@@ -501,18 +394,345 @@ impl BufferService {
+    async fn is_full(&self) -> Result<bool, BufferError> {
+        let current_size = self.estimate_spool_size().await?;
+        Ok(current_size as usize >= self.max_disk_size)
+    }
+
+    /// Returns `true` if the spool is empty, `false` otherwise.
+    async fn is_empty(&self) -> Result<bool, BufferError> {
+        let is_empty = sql::select_one()
+            .fetch_optional(&self.db)
+            .await
+            .map_err(BufferError::FetchFailed)?
+            .is_none();
+
+        Ok(is_empty)
+    }
+
+    /// Enqueues data into on-disk spool.
+    async fn enqueue(
+        &self,
+        key: QueueKey,
+        managed_envelope: ManagedEnvelope,
+    ) -> Result<(), BufferError> {
+        let received_at = managed_envelope.received_at().timestamp_millis();
+        sql::insert(
+            key,
+            managed_envelope.into_envelope().to_vec().unwrap(),
+            received_at,
+        )
+        .execute(&self.db)
+        .await
+        .map_err(BufferError::InsertFailed)?;
+
+        relay_statsd::metric!(counter(RelayCounters::BufferWrites) += 1);
+        Ok(())
+    }
+}
+
+/// The state which defines the [`BufferService`] behaviour.
+#[derive(Debug)]
+enum BufferState {
+    /// Only memory buffer is enabled.
+    Memory { ram: InMemory },
+
+    /// The memory buffer is used, but also disk spool is configured.
+    ///
+    /// The disk will be used when the memory limit will be hit.
+    MemoryFileStandby { ram: InMemory, disk: OnDisk },
+
+    /// THe disk is full and we use the memory to save incoming envelopes as much as we can. But
+    /// the disk still used to read from.
+    MemoryFileRead { ram: InMemory, disk: OnDisk },
+
+    /// Only disk used for read/write operations.
+    File { disk: OnDisk },
+}
+
+impl BufferState {
+    /// Creates the initial [`BufferState`] depending on the provided config.
+    ///
+    /// If the on-disk spool is configured, the `BufferState::MemoryDiskStandby` will be returned or
+    /// `BufferState::Memory` otherwise.
+    fn new(max_memory_size: usize, on_disk: Option<OnDisk>) -> Self {
+        let in_memory = InMemory {
+            buffer: BTreeMap::new(),
+            max_memory_size,
+            used_memory: 0,
+        };
+        match on_disk {
+            Some(on_disk) => Self::MemoryFileStandby {
+                ram: in_memory,
+                disk: on_disk,
+            },
+            None => Self::Memory { ram: in_memory },
+        }
+    }
+
+    /// Becomes a different state, depdending on the current state and the current conditions of
+    /// underlying spool.
+    async fn update(&mut self, config: &Arc<Config>) -> Result<(), BufferError> {
+        match self {
+            BufferState::Memory { .. } => (),
+            BufferState::MemoryFileStandby { ram, disk } => {
+                if ram.is_full() {
+                    disk.spool(&mut ram.buffer, &mut ram.used_memory).await?;
```

Comment:
nit: It would be nice if `update(self, config: ...)` consumed the state and returned a new state. Then you could pass `ram` by value into `spool`.

Added bonus: You might not even need the `to_` methods any longer, because you could pass the internals (in this case `disk`) into the new variant by value. E.g. in this case

`return BufferState::File { disk }`

But it would require implementing `Default` on `BufferState` so you can `std::mem::take` the whole thing, so I'm not sure if it's worth it.

Reply:
Yeah, in my first iteration, that's how it worked, i'm ok with returning this behaviour back.